# 简介

数据集提供了1700多位K-Pop（韩流）偶像的信息，包括他们的艺名、全名、韩文名、出生日期、身高、体重、出生地以及之前和其他团体的关联。K-Pop是全球最受欢迎和最具活力的音乐产业之一，这个数据集适合韩流产业或者进行偶像研究的数据分析人员。


变量含义：
- `Stage Name`：偶像被熟知的舞台上名字，即艺名
- `Full Name`：偶像的全名
- `Korean Name`：偶像的韩文名字
- `K Stage Name`：偶像的韩文舞台名字
- `Date of Birth`：偶像的出生日期
- `Group`：偶像所属的韩流团体
- `Debut`：偶像出道的日期
- `Company`：偶像所属娱乐公司
- `Country`：偶像的原籍国家
- `Second Country`：偶像的第二个国家（如果有的话）

# 读取数据

导入分析所需要的数据库，并通过pandas的read_csv函数，将原始数据文件`e_commerce.csv`里的数据内容解析为DataFrame，并赋值给变量`original_data`。

In [1]:
import pandas as pd

In [3]:
original_data=pd.read_csv("./kpop_idols.csv")
original_data.head(10)

Stage Name      Full Name Korean Name K Stage Name Date of Birth  \
0      2Soul  Kim Younghoon         김영훈           이솔    10/09/1997   
1        A.M  Seong Hyunwoo         성현우          에이엠    31/12/1996   
2        Ace  Jang Wooyoung         장우영          에이스    28/08/1992   
3       Aeji      Kwon Aeji         권애지           애지    25/10/1999   
4       AhIn       Lee Ahin         이아인           아인    27/09/1999   
5       Ahra        Go Ahra         고아라           아라    21/02/2001   
6     Ahyeon    Jung Ahyeon         정아현           아현    11/04/2007   
7     Ahyoon     Choi Subin         최수빈           아윤    23/10/2004   
8     Ahyoon    Shin Ahyoon         신아윤           아윤    24/09/2003   
9    Ahyoung    Cho Jayoung         조자영           아영    26/05/1991   

         Group       Debut       Company      Country Second Country  Height  \
0    7 O'clock  26/08/2014        Jungle  South Korea            NaN   172.0   
1    Limitless   9/07/2019           ONO  South Korea            NaN   181.0   
2          VAV  31/10/2015        A team  South Korea            NaN   177.0   
3     Hash Tag  11/10/2017           LUK  South Korea            NaN   163.0   
4     MOMOLAND   9/11/2016   Double Kick  South Korea            NaN   160.0   
5     Favorite   5/07/2017        Astory  South Korea            NaN     NaN   
6  BABYMONSTER   0/01/1900            YG  South Korea            NaN     NaN   
7     BOTOPASS  26/08/2020       WKS ENE  South Korea            NaN     NaN   
8   Queenz Eye  24/10/2022  Big Mountain  South Korea            NaN     NaN   
9   Dal Shabet   3/01/2011    Happy Face  South Korea            NaN     NaN   

   Weight Birthplace Other Group Former Group Gender  
0    55.0        NaN         NaN          NaN      M  
1    62.0        NaN         NaN          NaN      M  
2    63.0        NaN         NaN          NaN      M  
3     NaN      Daegu         NaN          NaN      F  
4    44.0      Wonju         NaN          NaN      F  
5     NaN      Yeosu         NaN          NaN      F  
6     NaN        NaN         NaN          NaN      F  
7     NaN        NaN         NaN          NaN      F  
8     NaN      Seoul         NaN          NaN      F  
9     NaN      Seoul         NaN          NaN      F

# 评估数据

在这一部分，我将基于上述已经建立好的`original_data`这个DateFrame所包含的数据进行评估。
评估主要从两个维度展开：
- **结构性的整齐度**：数据是否符合“每一列是一个变量，每一行是一个观察值，每一个单元格是一个值”这三个标准。
- **内容性的干净度**：数据中是否存在丢失数据、重复数据以及无效数据。

In [26]:
original_data.sample(10)

Stage Name     Full Name Korean Name K Stage Name Date of Birth  \
1719      Yubin    Gong Yubin         공유빈           유빈     3/02/2005   
580      Hyunuk    Han Hyunuk         한현욱           현욱    26/09/1994   
1112        New  Choi Chanhee         최찬희            뉴    26/04/1998   
287        Ella  Kim Kyungjoo         김경주           엘라    26/03/1998   
784        Judy     Kim Dahye         김다혜           주디    16/05/1995   
1166      Rinji   Hwang Rinji         황린지           린지     5/05/2006   
1493      Tzuyu    Chou Tzuyu        저우쯔위           쯔위    14/06/1999   
1286   Seungmin  Kim Seungmin         김승민           승민    22/09/2000   
1059       Miso    Jeon Jimin         전지민           미소    25/10/2000   
1216   Sangyeon  Lee Sangyeon         이상연           상연     4/11/1996   

           Group       Debut      Company      Country Second Country  Height  \
1719     tripleS  13/02/2023      MODHAUS  South Korea            NaN   163.0   
580        IN2IT  27/01/2016  Star Empire  South Korea            NaN   179.0   
1112    The Boyz  31/12/2017      Cre.Ker  South Korea            NaN   175.0   
287          NaN         NaN          NaN  South Korea            NaN     NaN   
784    Blackswan  16/10/2020           DR  South Korea            NaN     NaN   
1166        Pixy  24/02/2021       ALLART  South Korea            NaN   169.0   
1493       Twice  20/10/2015          JYP       Taiwan            NaN   169.0   
1286  Stray Kids  25/03/2018          JYP  South Korea            NaN     NaN   
1059   Dreamnote   7/11/2018          iME  South Korea            NaN   160.0   
1216    The Boyz  31/12/2017      Cre.Ker  South Korea            NaN   176.0   

      Weight Birthplace Other Group         Former Group Gender  
1719     NaN     Yongin         NaN                  NaN      F  
580     63.0        NaN         NaN                  NaN      M  
1112    51.0     Jeonju         NaN                  NaN      M  
287      NaN        NaN         NaN  Cherry Bullet| Pixy      F  
784      NaN        NaN         NaN                  NaN      F  
1166     NaN        NaN         NaN                  NaN      F  
1493     NaN     Tainan         NaN                  NaN      F  
1286     NaN      Seoul         NaN                  NaN      M  
1059    43.0        NaN         NaN                  NaN      F  
1216    61.0        NaN         NaN                  NaN      M

从抽样的10条数据来看 ，数据符合"每一列是一个变量，每一行是一个观察值，每一个单元格是一个值"的标准。具体来看，每行是一个爱豆的公开个人信息，每列是爱豆的具体信息，因此不构成结构上的问题。

### 评估数据的干净度

在众多数据中优先需要观察的是`Debut`,`Company`,`Group`,`Country`字段完整且一致，这是分析行业趋势的基石。

In [10]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1778 entries, 0 to 1777
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Stage Name      1778 non-null   object 
 1   Full Name       1769 non-null   object 
 2   Korean Name     1768 non-null   object 
 3   K Stage Name    1777 non-null   object 
 4   Date of Birth   1776 non-null   object 
 5   Group           1632 non-null   object 
 6   Debut           1632 non-null   object 
 7   Company         1632 non-null   object 
 8   Country         1778 non-null   object 
 9   Second Country  62 non-null     object 
 10  Height          836 non-null    float64
 11  Weight          566 non-null    float64
 12  Birthplace      834 non-null    object 
 13  Other Group     140 non-null    object 
 14  Former Group    264 non-null    object 
 15  Gender          1778 non-null   object 
dtypes: float64(2), object(14)
memory usage: 222.4+ KB


从输出结果来看，数据总共存在1778条观察值。根据对观察值的重要性要求来看，`Debut`,`Company`存在空缺值。此外，`Date of Birth`的数据类型应该为日期时间而非字符串，应当对数据进行格式转换。

#### 评估缺失数据

在了解`Debut`存在缺失值之后，根据条件提取缺失观察值。

In [11]:
original_data[original_data["Debut"].isnull()]

Stage Name      Full Name Korean Name K Stage Name Date of Birth Group  \
11        Ailee      Lee Yejin         이예진          에일리    30/05/1989   NaN   
12         Aini    Kim Heejung         김희정          아이니    13/07/1991   NaN   
16        AleXa       Kim Seri         김세리          알렉사     6/12/1996   NaN   
29        Arang     Son Mnjung         손민정           아랑     8/03/2000   NaN   
46          B.I     Kim Hanbin         김한빈          비아이    22/10/1996   NaN   
...         ...            ...         ...          ...           ...   ...   
1708   Youngmin   Lim Youngmin         임영민           영민    25/12/1995   NaN   
1714      Yubin      Cho Yubin         조유빈           유빈     9/10/1999   NaN   
1725    Yujeong    Kim Yujeong         김유정           유정    14/02/1992   NaN   
1740     Yulhee     Kim Yulhee         김율희           율희    27/11/1997   NaN   
1776       Z-UK  Jeong Jaewook         정재욱           지욱    27/01/1993   NaN   

     Debut Company      Country Second Country  Height  Weight Birthplace  \
11     NaN     NaN  South Korea            USA   165.0     NaN     Denver   
12     NaN     NaN  South Korea            NaN   163.0    44.0        NaN   
16     NaN     NaN          USA            NaN     NaN     NaN        NaN   
29     NaN     NaN  South Korea            NaN     NaN     NaN        NaN   
46     NaN     NaN  South Korea            NaN     NaN     NaN     Goyang   
...    ...     ...          ...            ...     ...     ...        ...   
1708   NaN     NaN  South Korea            NaN     NaN     NaN      Busan   
1714   NaN     NaN  South Korea            NaN   156.0    40.0        NaN   
1725   NaN     NaN  South Korea            NaN   161.0    42.0      Seoul   
1740   NaN     NaN  South Korea            NaN   166.0    52.0    Bucheon   
1776   NaN     NaN  South Korea            NaN   174.0    62.0        NaN   

     Other Group  Former Group Gender  
11           NaN           NaN      F  
12           NaN  Pink Fantasy      F  
16           NaN           NaN      F  
29           NaN     NeonPunch      F  
46           NaN          iKON      M  
...          ...           ...    ...  
1708         MXM         AB6IX      M  
1714         NaN  Pink Fantasy      F  
1725         NaN        LABOUM      F  
1740         NaN        LABOUM      F  
1776         NaN        Bigflo      M  

[146 rows x 16 columns]

存在146条爱豆个人信息中缺失`Debut`变量值。

从输出结果来看，这些缺失`Debut`的个人数据，同时也缺失`Company`和`Group`变量值。为了检验猜想，我们将增加筛选条件，查看是否存在`Debut`变量值缺失且`Company`与`Group`变量值没有缺失的情况

In [12]:
original_data[original_data["Debut"].isnull() & original_data["Company"].notnull() & original_data["Group"].notnull()]

Empty DataFrame
Columns: [Stage Name, Full Name, Korean Name, K Stage Name, Date of Birth, Group, Debut, Company, Country, Second Country, Height, Weight, Birthplace, Other Group, Former Group, Gender]
Index: []

筛选出来的结果为0条，说明缺失`Debut`变量值的数据，同时也不具备有效的`Company`和`Group`值。具体来说，`Debut`表示爱豆个人或者其所属团体的出道日期，而`Company`和`Group`分别指代爱豆所属的娱乐公司和所属偶像团体，都是进行后续偶像行业分析的重要变量。倘若它们同时存在缺失/无效，我们有理由认为该类数据无法提供有效的意义，因此这些数据可以在后期被删除。

`Former Group`也同样存在缺失值的情况，因此需要根据条件提取缺失观察值。

In [13]:
original_data[original_data["Former Group"].isnull()]

Stage Name      Full Name Korean Name K Stage Name Date of Birth  \
0         2Soul  Kim Younghoon         김영훈           이솔    10/09/1997   
1           A.M  Seong Hyunwoo         성현우          에이엠    31/12/1996   
2           Ace  Jang Wooyoung         장우영          에이스    28/08/1992   
3          Aeji      Kwon Aeji         권애지           애지    25/10/1999   
4          AhIn       Lee Ahin         이아인           아인    27/09/1999   
...         ...            ...         ...          ...           ...   
1772        Ziu    Park Heejun         박희준           지우    16/06/1997   
1773         ZN      Bae Jinye         배진예           지엔     9/06/1994   
1774        Zoa     Cho Hyewon         조혜원           조아    31/05/2005   
1775       Zuho       Bae Juho         백주호           주호     4/07/1996   
1777       Zuny      Kim Joomi         김주미           주니     8/12/1994   

             Group       Debut      Company      Country Second Country  \
0        7 O'clock  26/08/2014       Jungle  South Korea            NaN   
1        Limitless   9/07/2019          ONO  South Korea            NaN   
2              VAV  31/10/2015       A team  South Korea            NaN   
3         Hash Tag  11/10/2017          LUK  South Korea            NaN   
4         MOMOLAND   9/11/2016  Double Kick  South Korea            NaN   
...            ...         ...          ...          ...            ...   
1772           VAV  31/10/2015       A team  South Korea            NaN   
1773        LABOUM  27/08/2014           NH  South Korea            NaN   
1774       Weeekly  30/07/2020       Play M  South Korea            NaN   
1775           SF9   5/10/2016          FNC  South Korea            NaN   
1777  Ladies' Code   7/03/2013      Polaris  South Korea            NaN   

      Height  Weight Birthplace Other Group Former Group Gender  
0      172.0    55.0        NaN         NaN          NaN      M  
1      181.0    62.0        NaN         NaN          NaN      M  
2      177.0    63.0        NaN         NaN          NaN      M  
3      163.0     NaN      Daegu         NaN          NaN      F  
4      160.0    44.0      Wonju         NaN          NaN      F  
...      ...     ...        ...         ...          ...    ...  
1772   185.0    70.0        NaN         NaN          NaN      M  
1773   169.0    48.0    Bucheon       UNI.T          NaN      F  
1774   170.0     NaN        NaN         NaN          NaN      F  
1775     NaN     NaN        NaN         NaN          NaN      M  
1777     NaN     NaN    Gwangju         NaN          NaN      F  

[1514 rows x 16 columns]

`Former Group`表示爱豆之前出道的团体，即存在二次出道的情况，但这并不是分析爱豆行业发展趋势的必要变量，且从输出结果来看有些缺失该变量的信息仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

检查所有列完全重复的行

In [14]:
original_data[original_data.duplicated()]

Empty DataFrame
Columns: [Stage Name, Full Name, Korean Name, K Stage Name, Date of Birth, Group, Debut, Company, Country, Second Country, Height, Weight, Birthplace, Other Group, Former Group, Gender]
Index: []

由上述查重步骤可知，所有列中完全重复的行为0条，即不存在单个爱豆个人信息的重复，但考虑到`Stage Name`可能会存在重复的可能性，因此可以检查特定列的重复值。

In [15]:
original_data[original_data.duplicated(subset="Stage Name",keep=False)]

Stage Name       Full Name Korean Name K Stage Name Date of Birth  \
7        Ahyoon      Choi Subin         최수빈           아윤    23/10/2004   
8        Ahyoon     Shin Ahyoon         신아윤           아윤    24/09/2003   
17        Alice     Song Joohee         송주희          앨리스    21/03/1990   
18        Alice  Cheon Jaeyoung         천재영          앨리스    20/02/2002   
22         Andy   Lui Chunyeung        루이쯔양           앤디     6/10/1994   
...         ...             ...         ...          ...           ...   
1744       Yuna       Shin Yuna         신유나           유나     9/12/2003   
1747      Yunji       Kim Yunji         김윤지           윤지    26/08/1996   
1748      Yunji       Lee Yunji         이윤지           윤지     4/04/1992   
1758       Yuri         Jo Yuri         조유리           유리    22/10/2001   
1759       Yuri       Kwon Yuri        권유리            유리     5/12/1989   

            Group       Debut         Company      Country Second Country  \
7        BOTOPASS  26/08/2020         WKS ENE  South Korea            NaN   
8      Queenz Eye  24/10/2022    Big Mountain  South Korea            NaN   
17    Hello Venus   9/05/2012        Fantagio  South Korea            NaN   
18         TRACER   3/04/2022       Gleamedia  South Korea            NaN   
22      7 O'clock  26/08/2014          Jungle    Hong Kong            NaN   
...           ...         ...             ...          ...            ...   
1744         ITZY  12/02/2019             JYP  South Korea            NaN   
1747        ARIAZ  23/10/2019     Rising Star  South Korea            NaN   
1748     Playback  25/06/2015         Coridel  South Korea            NaN   
1758       IZ*ONE  29/10/2018  Off The Record  South Korea            NaN   
1759         SNSD   4/08/2021       Good Luck  South Korea            NaN   

      Height  Weight Birthplace Other Group Former Group Gender  
7        NaN     NaN        NaN         NaN          NaN      F  
8        NaN     NaN      Seoul         NaN          NaN      F  
17     166.0    47.0      Wonju         NaN          NaN      F  
18       NaN     NaN        NaN         NaN          NaN      F  
22       NaN     NaN        NaN         NaN          NaN      M  
...      ...     ...        ...         ...          ...    ...  
1744     NaN     NaN      Suwon         NaN          NaN      F  
1747     NaN     NaN        NaN         NaN          NaN      F  
1748     NaN     NaN        NaN         NaN          NaN      F  
1758     NaN     NaN      Busan         NaN          NaN      F  
1759   167.0    45.0     Goyang         NaN          NaN      F  

[520 rows x 16 columns]

由此可见存在520条爱豆个人数据中的`Stage Name`值有重复的情况。但由于韩国人名字存在高度重复率的可能性，即一个相同艺名可能对应的是不同的爱豆实例，因此可以保留该数据。

### 评估不一致数据

不一致数据可能存在于`Company`变量中，我们需要检查是否存在多个值指代同一个公司的情况。

In [20]:
original_data["Company"].value_counts()

SM                58
JYP               56
FNC               52
Starship          44
YG                41
                  ..
About              2
MBK| Turbo Co.     2
AO                 2
Hunus              1
Midas              1
Name: Company, Length: 180, dtype: int64

从变量值来看，猜想未能成立，`Country`值不存在数据不一致的问题。还需要再检查一下`Second Country`中是否存在不同的值指向一个国家的可能性。

In [22]:
original_data["Second Country"].value_counts()

South Korea           21
USA                   19
Australia              4
Canada                 3
Taiwan                 2
China                  2
Japan                  2
Argentina              1
Senegal                1
Hong Kong| Taiwan      1
Argentina|USA          1
United States          1
South Korea| China     1
Romania                1
Fiji                   1
Seoul                  1
Name: Second Country, dtype: int64

根据上述步骤可以知晓，`Second Country`变量值确实存在问题，主要集中在：
- 国家名称不一致，如`USA`和`United States`同时存在
- 组合值格式混乱，分隔符不一致
- 包含国家非国家实体，城市名混入国家列
- 国家和地区归属争议，如`Taiwan`和`HongKong`作为独立值出现，应该同时归属为`China`

为了确保`Country`,`Second Country`两个变量中的值格式一致，需要再查询`Country`中是否存在数据不一致的问题。

In [71]:
original_data["Country"].value_counts()

South Korea    1599
Japan            64
China            46
USA              18
Thailand         13
Taiwan            9
Canada            6
Hong Kong         5
Australia         4
Philippines       2
Indonesia         2
Vietnam           2
India             1
Germany           1
Belgium           1
Malaysia          1
New Zealand       1
Brazil            1
Russia            1
Macau             1
Name: Country, dtype: int64

根据输出内容可以发现存在表述上的值和`Second Country`不一致的问题，需要在后续进行以下调整：
- `USA`替换成`Unied States`
- `HongKong`,`Macau`,`Taiwan`替换成`China HongKong`,`China Macau`,`China Taiwan`

#### 评估无效或错误数据

采用`describe`方法，对数据统计信息首先快速了解

In [23]:
original_data.describe()

Height      Weight
count  836.000000  566.000000
mean   170.643541   57.224382
std      7.805094   16.831308
min    150.000000   38.000000
25%    164.000000   48.000000
50%    170.000000   57.000000
75%    177.000000   63.000000
max    190.000000  170.000000

从输出结果来看，`Weight`值存在异常数据，即最大值==170，或许会对后续数值分析产生影响。因此，我们优先筛选出`Weight`数值异常的观察值。

In [24]:
original_data[original_data["Weight"]==170]

Stage Name     Full Name Korean Name K Stage Name Date of Birth  \
290       Ellyn   Bang Sunhee         방선희           엘린    19/10/2002   
1239  Seokcheol  Lee Seokchul         이석철           석철    11/01/2000   

             Group       Debut    Company      Country Second Country  Height  \
290       Girlkind  16/01/2018  Nextlevel  South Korea            NaN     NaN   
1239  TheEastLight  15/11/2018   Stardium  South Korea            NaN     NaN   

      Weight Birthplace Other Group Former Group Gender  
290    170.0        NaN         NaN          NaN      F  
1239   170.0        NaN         NaN          NaN      M

从输出结果看，`Weight`变量为170的观察值存在`Height`变量有空缺值的情况。根据爱豆行业对身高体重的控制要求来看，体重不可能达到170的情况，因此有可能存在`Weight`值和`Height`值混淆的可能。
需要对数值进行调整。

但此猜想不一定准确，为了验证猜想，我们需要缩小筛选范围即`Weight`>100,并增加筛选条件，查看是否存在`Weight`变量大于100，且`Height`不为空值的值。

In [25]:
original_data[(original_data["Weight"]>100) & (original_data["Height"].notnull())]

Empty DataFrame
Columns: [Stage Name, Full Name, Korean Name, K Stage Name, Date of Birth, Group, Debut, Company, Country, Second Country, Height, Weight, Birthplace, Other Group, Former Group, Gender]
Index: []

根据输出结果，可以得知猜想正确，即爱豆体重不可能超过100，甚至连张兵（官方数据）都做不到。故此，极有可能存在`Weight`与`Height`两个变量值混淆的可能性，需要在后续步骤中进行清理。

# 清理数据

根据上述的数据评估得出的结论，我们需要进行的数据清理包括：
- `Date of Birth`变量的数据类型需要转换为日期时间。
- 把同步缺失`Debut`,`Company`,`Group`这三个变量的观察值删除。
- 把`Second Country`中的分割符进行统一，删除空格。
- 把`Second Country`变量值的`USA`替换成`United States`。
- 把`Second Country`变量值的`Seoul`替换成`South Korea`。
- 把`Second Country`变量值中的`Taiwan`和`HongKong`替换为`China Taiwan`和`China HongKong`。
- 把`Country`变量值中的`USA`替换成`United States`。
- 把`Country`变量值中的`Taiwan`和`HongKong`,`Macau`替换为`China Taiwan`和`China HongKong`,`China Macau`。
- 对`Weight`,`Height`中存在异常的数值进行调整。

In [28]:
cleaned_data=original_data.copy()
cleaned_data.head(10)

Stage Name      Full Name Korean Name K Stage Name Date of Birth  \
0      2Soul  Kim Younghoon         김영훈           이솔    10/09/1997   
1        A.M  Seong Hyunwoo         성현우          에이엠    31/12/1996   
2        Ace  Jang Wooyoung         장우영          에이스    28/08/1992   
3       Aeji      Kwon Aeji         권애지           애지    25/10/1999   
4       AhIn       Lee Ahin         이아인           아인    27/09/1999   
5       Ahra        Go Ahra         고아라           아라    21/02/2001   
6     Ahyeon    Jung Ahyeon         정아현           아현    11/04/2007   
7     Ahyoon     Choi Subin         최수빈           아윤    23/10/2004   
8     Ahyoon    Shin Ahyoon         신아윤           아윤    24/09/2003   
9    Ahyoung    Cho Jayoung         조자영           아영    26/05/1991   

         Group       Debut       Company      Country Second Country  Height  \
0    7 O'clock  26/08/2014        Jungle  South Korea            NaN   172.0   
1    Limitless   9/07/2019           ONO  South Korea            NaN   181.0   
2          VAV  31/10/2015        A team  South Korea            NaN   177.0   
3     Hash Tag  11/10/2017           LUK  South Korea            NaN   163.0   
4     MOMOLAND   9/11/2016   Double Kick  South Korea            NaN   160.0   
5     Favorite   5/07/2017        Astory  South Korea            NaN     NaN   
6  BABYMONSTER   0/01/1900            YG  South Korea            NaN     NaN   
7     BOTOPASS  26/08/2020       WKS ENE  South Korea            NaN     NaN   
8   Queenz Eye  24/10/2022  Big Mountain  South Korea            NaN     NaN   
9   Dal Shabet   3/01/2011    Happy Face  South Korea            NaN     NaN   

   Weight Birthplace Other Group Former Group Gender  
0    55.0        NaN         NaN          NaN      M  
1    62.0        NaN         NaN          NaN      M  
2    63.0        NaN         NaN          NaN      M  
3     NaN      Daegu         NaN          NaN      F  
4    44.0      Wonju         NaN          NaN      F  
5     NaN      Yeosu         NaN          NaN      F  
6     NaN        NaN         NaN          NaN      F  
7     NaN        NaN         NaN          NaN      F  
8     NaN      Seoul         NaN          NaN      F  
9     NaN      Seoul         NaN          NaN      F

把`Date of Birth`变量的数据类型转换为日期时间

In [29]:
cleaned_data["Date of Birth"]=pd.to_datetime(cleaned_data["Date of Birth"])
cleaned_data["Date of Birth"]

C:\Users\kris2\AppData\Local\Temp\ipykernel_11860\1337323363.py:1: UserWarning: Parsing '31/12/1996' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cleaned_data["Date of Birth"]=pd.to_datetime(cleaned_data["Date of Birth"])
C:\Users\kris2\AppData\Local\Temp\ipykernel_11860\1337323363.py:1: UserWarning: Parsing '28/08/1992' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cleaned_data["Date of Birth"]=pd.to_datetime(cleaned_data["Date of Birth"])
C:\Users\kris2\AppData\Local\Temp\ipykernel_11860\1337323363.py:1: UserWarning: Parsing '25/10/1999' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cleaned_data["Date of Birth"]=pd.to_datetime(cleaned_data["Date of Birth"])
C:\Users\kris2\AppData\Local\Temp\ipykernel_11860\1337323363.py:1: UserWarning: Parsing '27/09/1999' in DD/MM/YYYY format. Provide format or specify infer_datetime_fo

0      1997-10-09
1      1996-12-31
2      1992-08-28
3      1999-10-25
4      1999-09-27
          ...    
1773   1994-09-06
1774   2005-05-31
1775   1996-04-07
1776   1993-01-27
1777   1994-08-12
Name: Date of Birth, Length: 1778, dtype: datetime64[ns]

把同步缺失`Debut`,`Company`,`Group`这三个变量的观察值删除,并查看删除后该列空缺值的个数和。

In [31]:
cleaned_data.dropna(subset=["Debut"],inplace=True)

In [32]:
cleaned_data["Debut"].isnull().sum()

0

把`Second Country`变量值的`USA`替换成`United States`。

把`Second Country`变量值的`Seoul`替换成`South Korea`。

并且检查替换后的两个值的数量。

In [46]:
cleaned_data["Second Country"]=cleaned_data["Second Country"].replace({"USA":"United States","Seoul":"South Korea"})
cleaned_data

Stage Name      Full Name Korean Name K Stage Name Date of Birth  \
0         2Soul  Kim Younghoon         김영훈           이솔    1997-10-09   
1           A.M  Seong Hyunwoo         성현우          에이엠    1996-12-31   
2           Ace  Jang Wooyoung         장우영          에이스    1992-08-28   
3          Aeji      Kwon Aeji         권애지           애지    1999-10-25   
4          AhIn       Lee Ahin         이아인           아인    1999-09-27   
...         ...            ...         ...          ...           ...   
1772        Ziu    Park Heejun         박희준           지우    1997-06-16   
1773         ZN      Bae Jinye         배진예           지엔    1994-09-06   
1774        Zoa     Cho Hyewon         조혜원           조아    2005-05-31   
1775       Zuho       Bae Juho         백주호           주호    1996-04-07   
1777       Zuny      Kim Joomi         김주미           주니    1994-08-12   

             Group       Debut      Company      Country Second Country  \
0        7 O'clock  26/08/2014       Jungle  South Korea            NaN   
1        Limitless   9/07/2019          ONO  South Korea            NaN   
2              VAV  31/10/2015       A team  South Korea            NaN   
3         Hash Tag  11/10/2017          LUK  South Korea            NaN   
4         MOMOLAND   9/11/2016  Double Kick  South Korea            NaN   
...            ...         ...          ...          ...            ...   
1772           VAV  31/10/2015       A team  South Korea            NaN   
1773        LABOUM  27/08/2014           NH  South Korea            NaN   
1774       Weeekly  30/07/2020       Play M  South Korea            NaN   
1775           SF9   5/10/2016          FNC  South Korea            NaN   
1777  Ladies' Code   7/03/2013      Polaris  South Korea            NaN   

      Height  Weight Birthplace Other Group Former Group Gender  
0      172.0    55.0        NaN         NaN          NaN      M  
1      181.0    62.0        NaN         NaN          NaN      M  
2      177.0    63.0        NaN         NaN          NaN      M  
3      163.0     NaN      Daegu         NaN          NaN      F  
4      160.0    44.0      Wonju         NaN          NaN      F  
...      ...     ...        ...         ...          ...    ...  
1772   185.0    70.0        NaN         NaN          NaN      M  
1773   169.0    48.0    Bucheon       UNI.T          NaN      F  
1774   170.0     NaN        NaN         NaN          NaN      F  
1775     NaN     NaN        NaN         NaN          NaN      M  
1777     NaN     NaN    Gwangju         NaN          NaN      F  

[1632 rows x 16 columns]

In [53]:
len(cleaned_data[cleaned_data["Second Country"] =="USA"])

0

In [55]:
len(cleaned_data[cleaned_data["Second Country"]=="Soeul"])

0

把`Second Country`变量值中的`Taiwan`和`HongKong`替换为`China Taiwan`和`China HongKong`。

In [50]:
cleaned_data["Second Country"]=cleaned_data["Second Country"].replace({"Taiwan":"China Taiwan","HongKong":"China HongKong"})

In [52]:
len(cleaned_data[cleaned_data["Second Country"]=="Taiwan"])

0

对`Weight`,`Height`中存在异常的数值进行调整,使其数值互换，形成合理的数据。

In [56]:
mask=(cleaned_data["Weight"]==170) & (cleaned_data["Height"].isnull())
affected_rows=cleaned_data[mask]

In [59]:
cleaned_data.loc[mask,"Height"]=cleaned_data.loc[mask,"Weight"]
cleaned_data.loc[mask,"Weight"]=pd.NA

In [60]:
print(cleaned_data.loc[mask,["Stage Name","Height","Weight"]])

     Stage Name  Height Weight
290       Ellyn   170.0   <NA>
1239  Seokcheol   170.0   <NA>


In [63]:
cleaned_data["Second Country"]=cleaned_data["Second Country"].str.replace(r"\s*\|\s*","|",regex=True)

In [64]:
cleaned_data["Second Country"].value_counts()

South Korea         21
United States       17
Australia            4
China Taiwan         2
Canada               2
Japan                2
Argentina            1
Senegal              1
Hong KongTaiwan      1
ArgentinaUSA         1
South KoreaChina     1
Romania              1
Fiji                 1
China                1
Name: Second Country, dtype: int64

把`Country`变量值中的`USA`替换成`United States`。

In [72]:
cleaned_data["Country"]=cleaned_data["Country"].replace({"USA":"United States","Taiwan":"China Taiwan","Macau":"China Macau","HongKong":"China HongKong"})

In [74]:
print(len(cleaned_data[cleaned_data["Country"]=="USA"]))
print(len(cleaned_data[cleaned_data["Country"]=="Taiwan"]))
print(len(cleaned_data[cleaned_data["Country"]=="Macau"]))
print(len(cleaned_data[cleaned_data["Country"]=="HongKong"]))

0
0
0
0


# 保存清理好的数据

完成数据的清理之后，把干净整齐的数据保存到新的文件中，文件名为`kpop_idols_cleaned.csv`

In [77]:
cleaned_data.to_csv("kpop_idols_cleaned.csv",index=False)

In [81]:
pd.read_csv("kpop_idols_cleaned.csv").sample(15)

Stage Name     Full Name Korean Name K Stage Name Date of Birth  \
1228      Soeun    Bang Soeun         방소은           소은    2005-10-12   
1233     Sohyun   Kwon Sohyun         권소현           소현    1994-08-30   
379      Haneul    Lee Haneul         이하늘           하늘    1996-03-06   
1452      Yebin    Baek Yebin         백예빈           예빈    1997-07-13   
1580    Yugyeom   Kim Yugyeom         김유겸           유겸    1997-11-17   
727      Jun. K    Kim Minjun         김민준          준케이    1988-01-15   
1257       Soso  Wang Chingyi         왕징이           소소    2001-03-14   
911        Mill   Lee Yongsoo         이용수            밀    1999-03-30   
842      Layeon     Lee Jimin         이지민           라연    1998-08-13   
850     Lee Soo      Jin Isoo         진이수           이수    2001-06-27   
1619      Yuvin    Song Yuvin         송유빈           유빈    1998-04-28   
557      J.seph  Kim Taehyung         김태형          제이셉    1992-06-21   
1317    Sunyoul     Seon Yein         선예인           선율    1996-06-11   
1247     Soobin   Choi Soobin         최수빈           수빈    2000-05-12   
1631       Zuny     Kim Joomi         김주미           주니    1994-08-12   

             Group       Debut      Company       Country Second Country  \
1228        TRI.BE  17/02/2021  TR - Mellow   South Korea            NaN   
1233       4minute  15/06/2009         Cube   South Korea            NaN   
379          ARGON  11/03/2019          MSH   South Korea            NaN   
1452           DIA  14/07/2015          MBK   South Korea            NaN   
1580          GOT7  16/01/2014          JYP   South Korea            NaN   
727            2PM   4/07/2008          JYP   South Korea            NaN   
1257          GWSN   5/09/2018     Kiwi Pop  China Taiwan            NaN   
911      OnlyOneOf  28/05/2019         RSVP   South Korea            NaN   
842     Purplebeck  24/06/2019      Majesty   South Korea            NaN   
850       PRECIOUS  23/09/2020          Umi   South Korea            NaN   
1619         B.O.Y  20/01/2020  Music Works   South Korea            NaN   
557           KARD  29/03/2007          DSP   South Korea  United States   
1317      UP10TION  10/09/2015    TOP Media   South Korea            NaN   
1247           TXT   4/03/2019      Big Hit   South Korea            NaN   
1631  Ladies' Code   7/03/2013      Polaris   South Korea            NaN   

      Height  Weight Birthplace Other Group Former Group Gender  
1228     NaN     NaN        NaN         NaN          NaN      F  
1233   162.0     NaN      Seoul         NaN          NaN      F  
379    175.0     NaN        NaN         NaN          NaN      M  
1452   164.0    43.0  Chuncheon       UNI.T          NaN      F  
1580   180.0    64.0  Namyangju         NaN          NaN      M  
727    180.0    68.0      Daegu         NaN          NaN      M  
1257     NaN     NaN        NaN         NaN          NaN      F  
911      NaN     NaN        NaN         NaN          NaN      M  
842    170.0    51.0        NaN         NaN          NaN      F  
850    163.0    47.0    Incheon         NaN          NaN      F  
1619     NaN     NaN        NaN         NaN       MYTEEN      M  
557      NaN     NaN        NaN         NaN          NaN      M  
1317     NaN     NaN      Wando         NaN          NaN      M  
1247   184.0    65.0        NaN         NaN          NaN      M  
1631     NaN     NaN    Gwangju         NaN          NaN      F